controllers/usersignup/usersignup_controller.go

```diff
@@ -357,7 +357,9 @@ func (r *Reconciler) checkIfMurAlreadyExists(reqLogger logr.Logger, config toolc
 
 			// if the Space is not Ready, another reconcile will occur when space status is updated since this controller watches space (without a predicate)
 			reqLogger.Info("Checking whether Space is Ready", "Space", space.Name)
-			if !condition.IsTrueWithReason(space.Status.Conditions, toolchainv1alpha1.ConditionReady, toolchainv1alpha1.SpaceProvisionedReason) {
+			if !(condition.IsTrueWithReason(space.Status.Conditions, toolchainv1alpha1.ConditionReady, toolchainv1alpha1.SpaceProvisionedReason) ||
+				// don't mark the user signup as incomplete if the space is updating - wait for the space to finish first.
+				condition.IsFalseWithReason(space.Status.Conditions, toolchainv1alpha1.ConditionReady, toolchainv1alpha1.SpaceUpdatingReason)) {
 				return true, r.updateIncompleteStatus(userSignup, fmt.Sprintf("space %s was not ready", space.Name))
 			}
```

Comment:
I wonder if it's safer to not update any conditions at all if the space in in the updating state. Something like this:
```
if condition.IsFalseWithReason(space.Status.Conditions, toolchainv1alpha1.ConditionReady, toolchainv1alpha1.SpaceUpdatingReason) {
	// don't mark the user signup as incomplete if the space is updating - wait for the space to get ready first.
	return true, nil
}
if !condition.IsTrueWithReason(space.Status.Conditions, toolchainv1alpha1.ConditionReady, toolchainv1alpha1.SpaceProvisionedReason {
	return true, r.updateIncompleteStatus(userSignup, fmt.Sprintf("space %s was not ready", space.Name))
}
```
Basically if the space is being updated then we don't touch the usersignup conditions. It stays in whatever status it's already in. Trying to avoid a potential edge case situation when the UserSignup in some incomplete state already (some error state for example) and the space is being updated. We should not (re-)set the UserSignup complete condition to true just because space is being updated. We just should avoid setting already "complete" condition to "incomplete" when updating. I'm not sure if it can happen... But just to be on the safe side.

@MatousJobanek, @mfrancisc WDYT?

And if we do want to cover that use case then we will need to add a unit test for this too.

Reply:
If the UserSignup is in an incomplete (broken) state already, then reflecting the status from the Space doesn't change on the UserSignup state much. It either eventually recover after the Space update or not. In that case, the state of the UserSignup should be still incomplete.

TBH, I wouldn't make the logic hard-coded so it would rely on the specific reason in the Space CR. As soon as the UserSignup ends up in a complete state (after creating MUR and Space CRs and waiting for their readiness), then I wouldn't reflect any updates in status of the Space in the UserSignup CR. The job of the UserSignup controller is done - it's complete and it should stay as it is no matter what is going on with the actual Space CR. This is also the reason why we named the condition as "complete" and not "ready"

1. UserSignup is approved
2. It creates MUR, Space, and SpaceBinding
3. It waits until Space is ready
4. When Space is ready then it marks the "complete" condition to true
5. If there is something going on with the Space CR later on, then the UserSignup should care about that - the "Complete" condition should stay "true"

Reply:
I like @MatousJobanek proposal to not reflect the Space status in the UserSignup once the UserSignup is "Complete" 👍

Reply:
But maybe we will need a way to notify the user in case there is an issue with the space he/she is trying to access.  I know a User might have access to different (work)spaces see RHTAP for example.

Reply:
Just to make sure I'm understanding this correctly: we want to keep usersignups that have already completed, but mark all the other signups as incomplete (provided they're not already rejected for another reason, such as a deactivated/banned user).

Reply:
If I understand @MatousJobanek the proposal is to stop setting the `complete` condition to `false` if the Space is not ready. I like the idea in general. However I have a question.
There are two options. We just do not set the `complete` condition at all until the space is ready. So the condition could be missing/unknown until space is ready.
Or we do something like this:
```
if !condition.IsTrueWithReason(space.Status.Conditions, toolchainv1alpha1.ConditionReady, toolchainv1alpha1.SpaceProvisionedReason && !USER_SIGNUP_COMPLETE_IS_TRUE {
	return true, r.updateIncompleteStatus(userSignup, fmt.Sprintf("space %s is not ready", space.Name))
}
```
So, it does set it to incomplete with the `space nor ready` message but only if it's not complete yet. Just to provide more details why it's not complete yet. If it was set to compelete then we don't change it no matter what status of the space.

The other cases (banned, etc) stay intact.

Reply:
I've implemented the second option with 97d45fe, though I can change it if we want to go with the first.

```go
if !(condition.IsTrueWithReason(space.Status.Conditions, toolchainv1alpha1.ConditionReady, toolchainv1alpha1.SpaceProvisionedReason) ||
	condition.IsTrue(userSignup.Status.Conditions, toolchainv1alpha1.UserSignupComplete)) {
	return true, r.updateIncompleteStatus(userSignup, fmt.Sprintf("space %s was not ready", space.Name))
}
```

Reply:
> But maybe we will need a way to notify the user in case there is an issue with the space he/she is trying to access. 

let's address this as a separate task. Such a logic could be placed either here or for example in registration-service (we could provide similar functionality for (work)space-lister too).
For the DevSandbox use-case, we would need some support on UI side for that too, to show more states or errors of the UserSignup resource.

> So, it does set it to incomplete with the space nor ready message but only if it's not complete yet. Just to provide more details why it's not complete yet. If it was set to compelete then we don't change it no matter what status of the space.

Yes, this is exactly what I meant. Let me update the flow by adding another line/step
1. UserSignup is approved
2. It creates MUR, Space, and SpaceBinding
3. It waits until Space is ready
4. When UserSignup controller gets reconciled and the Space is still not ready, then it updates the condition with `space <space_name> is not ready`
5. When Space is ready then it marks the "complete" condition to true
6. If there is something going on with the Space CR later on, then the UserSignup shouldn't care about that - the "Complete" condition should stay "true" 

so the UserSignup controller reflects that the Space is not ready, but only until it gets ready for the first time. The it marks the condition as complete and whatever happens to the Space is ignored by the controller

> The other cases (banned, etc) stay intact.

yes, exactly

Reply:
in other words, exactly as @sadlerap did in his latest changes :pray: